include/ginkgo/core/solver/chebyshev.hpp

```diff
@@ -0,0 +1,236 @@
+        /**
+         * The pair of foci of ellipse, which covers the eigenvalues of
+         * preconditioned system. It is usually be {lower bound of eigval, upper
+         * bound of eigval} of preconditioned real matrices.
+         */
+        std::pair<value_type, value_type> GKO_FACTORY_PARAMETER_VECTOR(
+            foci, value_type{0}, value_type{1});
+
+        /**
+         * Default initial guess mode. The available options are under
+         * initial_guess_mode.
+         */
+        initial_guess_mode GKO_FACTORY_PARAMETER_SCALAR(
+            default_initial_guess, initial_guess_mode::provided);
+    };
+
+    GKO_ENABLE_LIN_OP_FACTORY(Chebyshev, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+    /**
+     * Create the parameters from the property_tree.
+     * Because this is directly tied to the specific type, the value/index type
+     * settings within config are ignored and type_descriptor is only used
+     * for children configs.
+     *
+     * @param config  the property tree for setting
+     * @param context  the registry
+     * @param td_for_child  the type descriptor for children configs. The
+     *                      default uses the value type of this class.
+     *
+     * @return parameters
+     */
+    static parameters_type parse(const config::pnode& config,
+                                 const config::registry& context,
+                                 const config::type_descriptor& td_for_child =
+                                     config::make_type_descriptor<ValueType>());
+
+protected:
+    void apply_impl(const LinOp* b, LinOp* x) const override;
+
+    template <typename VectorType>
+    void apply_dense_impl(const VectorType* b, VectorType* x,
+                          initial_guess_mode guess) const;
+
+    void apply_impl(const LinOp* alpha, const LinOp* b, const LinOp* beta,
+                    LinOp* x) const override;
+
+    void apply_with_initial_guess_impl(const LinOp* b, LinOp* x,
+                                       initial_guess_mode guess) const override;
+
+    void apply_with_initial_guess_impl(const LinOp* alpha, const LinOp* b,
+                                       const LinOp* beta, LinOp* x,
+                                       initial_guess_mode guess) const override;
+
+    void set_relaxation_factor(
+        std::shared_ptr<const matrix::Dense<ValueType>> new_factor);
+
+    explicit Chebyshev(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Chebyshev>(std::move(exec))
+    {}
+
+    explicit Chebyshev(const Factory* factory,
+                       std::shared_ptr<const LinOp> system_matrix);
+
+private:
+    std::shared_ptr<const LinOp> solver_{};
+    // num_generated_scalar_ tracks the number of generated scalar alpha
+    // and beta.
+    mutable size_type num_generated_scalar_ = 0;
+    // num_max_generation_ is the number of generated scalar kept in the
+    // workspace.
+    mutable size_type num_max_generation_ = 3;
```

Comment:
I'm a bit wary of mutable members beyond well-controlled cases like the workspace. They might make future work on thread-safety more complicated. Not that I have a good alternative for it, but I want to bring it up.

Reply:
Isn't wokrspace handled in similar way?

Reply:
We don't have any mutability in our LinOps right now beyond vector caches and workspaces, most of the `mutable` uses are in loggers etc. So the mutability is constrained into a single class. That's not the case here.

In general, if I write code, I expect a `const` object to not change internally, and the only reasons we don't follow this requirement fully right now seems to be the need for cache data structures to avoid reallocation, and the fact that loggers are `const` and can't store data themselves.